mindsdb/integrations/handlers/langchain_handler/langchain_handler.py

```diff
@@ -269,12 +274,17 @@ def default_completion(self, df, args=None, pred_args=None):
         pred_args['tools'] = args.get('tools') if 'tools' not in pred_args else pred_args.get('tools', [])
 
         # langchain tool setup
+        llm = self._create_chat_model(args, pred_args)
+        max_tokens = pred_args.get('max_tokens', args.get('max_tokens', self.default_max_tokens))
         model_kwargs = self._get_chat_model_params(args, pred_args)
-        tools = self._setup_tools(model_kwargs, pred_args, args['executor'])
+        tools = setup_tools(llm,
+                            model_kwargs,
+                            pred_args,
+                            args['executor'],
+                            self.default_agent_tools,
+                            self.write_privileges)
 
         # langchain agent setup
-        llm = self._create_chat_model(args, pred_args)
-        max_tokens = pred_args.get('max_tokens', args.get('max_tokens', self.default_max_tokens))
         memory = ConversationSummaryBufferMemory(llm=llm, max_token_limit=max_tokens)
         agent_name = pred_args.get('agent_name', args.get('agent_name', self.default_agent_model))
```

Comment:
nit, instead of this could be worth using `pydantic`. It can be simpler and less verbose overall once implemented. see here for an example [here](https://github.com/mindsdb/mindsdb/blob/c0151818e017cf72194e114db8f082d5ae78db99/mindsdb/integrations/handlers/rag_handler/settings.py#L281)

Just an idea, doesn't block merge.

May worth a quick chat offline to come up with a common approach for handlers working with user defined params which is more robust

Reply:
Turned into issue, as I agree but it is broader than this particular fix. #7497.